casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala

```diff
@@ -99,149 +172,236 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
-    "(4!) C!"   -> VOLATILE_EVENT             -> merges(F1 | S1, C_, S0),
-    "(C!) C!"   -> PERSISTENT_COULD_NOT_MATCH -> conflicts(C_ | S0, C_, S0),
-    "4! (4!)"   -> VOLATILE_EVENT             -> merges(F_, F1 | S1, S0),
-    "4! (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> conflicts(F_, C_ | S0, S0),
-    "4X (4!)"   -> VOLATILE_EVENT             -> merges(F_, F_ | S0, Nil),
-    "4X (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> merges(F_, C_ | S0, Nil),
-    "C! (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> merges(C_, C1 | S1, S0),
-    "CX (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> merges(C_, C_ | S0, Nil),
-    "(!4) 4!"   -> VOLATILE_EVENT             -> merges(S1 | F1, F_, S0),
-    "(!4) !C"   -> VOLATILE_EVENT             -> merges(S1 | F1, S0, C0),
-    "(!4) C!"   -> VOLATILE_EVENT             -> merges(S0 | F0, C_, S1),
-    "(4!) CX"   -> VOLATILE_EVENT             -> merges(F_ | S0, C_, Nil),
-    "(C!) (C!)" -> PERSISTENT_COULD_NOT_MATCH -> merges(C_ | S0, C_ | S0, Nil)
+    "!X !X"     -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, S0, Nil),
+    "!X !4"     -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, S1, F1),
+    "!X (!4)"   -> VolatileEvent.desc           -> VolatileEvent.mergeable(S0)(S0, F_)(Nil),
+    "!X !C"     -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, S1, C1),
+    "!X (!C)"   -> PersistentCouldMatch.desc    -> PersistentCouldMatch.conflicting(S0)(S0, C_)(Nil),
+    "!X 4X"     -> IncomingCouldMatch.desc      -> IncomingCouldMatch.conflicting(S0, F_, Nil),
+    "!X 4!"     -> HadItsMatch.desc             -> HadItsMatch.mergeable(S0, F_, S0),
+    "!X (4!)"   -> VolatileEvent.desc           -> coveredBy("!X (!4)"),
+    "!X 4!!"    -> HadItsMatch.desc             -> HadItsMatch.mergeable(S0, F_, R0),
+    "!X (4!!)"  -> PersistentNoMatch.desc       -> PersistentNoMatch.mergeable(S0)(F_, R0)(Nil),
+    "!X !!X"    -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, R0, Nil),
+    "!X !!4"    -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, R1, F1),
+    "!X (!!4)"  -> PersistentNoMatch.desc       -> coveredBy("!X (4!!)"),
+    "!X CX"     -> IncomingCouldMatch.desc      -> IncomingCouldMatch.conflicting(S0, C_, Nil),
+    "!X C!"     -> IncomingCouldMatch.desc      -> IncomingCouldMatch.conflicting(S0, C_, S0),
+    "!X (C!)"   -> PersistentCouldMatch.desc    -> PersistentCouldMatch.conflicting(S0)(C_, S0)(Nil),
+    "!4 !4"     -> CouldMatchSame.descConflicts -> CouldMatchSame.conflicting(S0, S1, F_),
+    "!4 !4"     -> CouldMatchSame.descMerges    -> CouldMatchSame.mergeable(S0, S1)(F0, F1),
+    "!4 (!4)"   -> VolatileEvent.desc           -> VolatileEvent.mergeable(S0)(S1, F_)(F0),
+    "(!4) (!4)" -> VolatileEvent.desc           -> VolatileEvent.mergeable(S0, F_)(S0, F_)(Nil),
+    "!4 !C"     -> CouldMatchSame.descMerges    -> CouldMatchSame.mergeable(S0, S1)(F0, C1),
```

Comment:
The number of variants of arities is confusing and inconsistent across the cases.
E.g. here it should say `CouldMatchSame(S0)(S1)(F0, C1)` to be consistent
 
Let's have all MergeabilityCases take 3 curried parameters: left, right, base.
The last one can be vararg.

For 'volatile/local' cases we can have all of them be wildcards.

```scala
def samePolarityMerge(left: Rho)(right: Rho)(base: Rho*): MergrabilityTestCase = new ...

def volatileEvent(lefts: Rho*)(rights: Rho*)(base: Rho*): MergrabilityTestCase = new ...
```

This works with an `apply` method on a companion object as well.

Assertions underneath then become
e.g. 
```
//all are linear
assert((lefts ++ rights).forall(_.cardinality == Linear))
//same polarity
assert(left.polarity == right.polarity) 
//different polarities within one side 
assert(noDuplicatesBy(lefts)(_.polarity))
```

Reply:
Good idea, I like it